File: src/app/services/trait-options.service.ts
import { Injectable } from '@angular/core';

interface MouthOption {name: string; selected: boolean;}
interface HeadCoveringOption {name: string; selected: boolean;}
interface EyesOption {name: string; selected: boolean;}
interface ClothesOption {name: string; selected: boolean;}
interface FurOption {name: string; selected: boolean;}
interface BackgroundOption {name: string; selected: boolean;}
interface TranscendenceOption {name: string; selected: boolean;}

@Injectable({
  providedIn: 'root'
})
export class TraitOptionsService {

  headCoveringOptions: HeadCoveringOption[] = [
    { name: 'Horns', selected: false },
    { name: 'Angel Halo', selected: false },
    { name: 'Golden Tiara', selected: false },
    { name: 'Toxic Liquid', selected: false },
    { name: 'Robotic Mask', selected: false },
    { name: 'Engraved Mask', selected: false },
    { name: 'Golden Kabuto', selected: false },
    { name: 'Pilot Helmet', selected: false },
    { name: 'Suit Hat', selected: false },
    { name: 'Crown', selected: false },
    { name: 'Ripple Of Magic', selected: false },
    { name: 'Straw Hat', selected: false },
    { name: 'Ninja Bandana', selected: false },
    { name: 'Headphones', selected: false },
    { name: 'Bob', selected: false },
    { name: 'Earring', selected: false },
  ];

  eyesOptions: EyesOption[] = [
    { name: 'Bloodshot', selected: false },
    { name: 'Scarlet', selected: false },
    { name: 'Golden Patterns', selected: false },
    { name: 'Toxic Green', selected: false },
    { name: 'Robotic', selected: false },
    { name: 'Striped', selected: false },
    { name: 'Wisdom (Closed)', selected: false },
    { name: 'Blindfolded', selected: false },
    { name: 'Sunglasses', selected: false },
    { name: 'Dollars', selected: false },
    { name: 'Sharingan', selected: false },
    { name: 'Happy (Closed)', selected: false },
    { name: 'Bored', selected: false },
    { name: 'Disdainful', selected: false },
    { name: 'Hypnotized', selected: false },
    { name: 'Normal', selected: false },
  ];

  mouthOptions: MouthOption[] = [
    { name: 'Magma Fangs', selected: false },
    { name: 'Blue Psyche', selected: false },
    { name: 'Gold Mustache', selected: false },
    { name: 'Gas Mask', selected: false },
    { name: 'Robot Mask', selected: false },
    { name: 'Engraved Mask', selected: false },
    { name: 'Gold Teeth', selected: false },
    { name: 'Twig Nibbles', selected: false },
    { name: 'Smoking', selected: false },
    { name: 'Tongue Out', selected: false },
    { name: 'Disgust', selected: false },
    { name: 'Rabbit', selected: false },
    { name: 'Hate', selected: false },
    { name: 'Happy', selected: false },
    { name: 'Confused', selected: false },
    { name: 'Neutral', selected: false },
  ];

  clothesOptions: ClothesOption[] = [
    { name: 'Winged Dark Armor', selected: false },
    { name: 'Winged Golden Armor', selected: false },
    { name: 'Ornamental Armor', selected: false },
    { name: 'Military', selected: false },
    { name: 'Robot', selected: false },
    { name: 'Samurai', selected: false },
    { name: 'Shoulder Cloak', selected: false },
    { name: 'Pilot Suit', selected: false },
    { name: 'Suit', selected: false },
    { name: "King's Cape", selected: false },
    { name: 'Black Jacket', selected: false },
    { name: 'Overalls', selected: false },
    { name: 'Blue Coat', selected: false },
    { name: 'Scarf', selected: false },
    { name: 'Polo Shirt', selected: false },
    { name: 'Orange Jacket', selected: false },
  ];

  furOptions: FurOption[] = [
    { name: 'Magma', selected: false },
    { name: 'Angel', selected: false },
    { name: 'Golden Obsidian', selected: false },
    { name: 'Dirty', selected: false },
    { name: 'Robot', selected: false },
    { name: 'Scarred', selected: false },
    { name: 'Beige', selected: false },
    { name: 'Dark Brown', selected: false },
    { name: 'Light Brown', selected: false },
    { name: 'Orange', selected: false },
  ];

  backgroundOptions: BackgroundOption[] = [
    { name: 'Hell', selected: false },
    { name: 'Heavenly Skies', selected: false },
    { name: 'Ancient Egypt', selected: false },
    { name: 'Green Gradient', selected: false },
    { name: 'Turquoise Gradient', selected: false },
    { name: 'Red Gradient', selected: false },
    { name: 'Soft Beige', selected: false },
    { name: 'Blue Grey', selected: false },
    { name: 'Dark Blue', selected: false },
    { name: 'Red', selected: false },
    { name: 'Soft Mauve', selected: false },
    { name: 'Light Blue', selected: false },
    { name: 'Orange', selected: false },
  ];
  
  transcendenceOptions: TranscendenceOption[] = [
    { name: 'Electrified', selected: false }, //0
    { name: 'Neon Nose', selected: false }, //1
  ];


  constructor() {}

  getTraitIndex(trait: string, type: string) {
    let options: any[] = [];
    switch (type.toLowerCase()) {
      case 'head covering':
        options = this.headCoveringOptions;
        break;
      case 'eyes':
        options = this.eyesOptions;
        break;
      case 'mouth':
        options = this.mouthOptions;
        break;
      case 'clothes':
        options = this.clothesOptions;
        break;
      case 'fur':
        options = this.furOptions;
        break;
      case 'background':
        options = this.backgroundOptions;
        break;
      case 'transcendence':
        options = this.transcendenceOptions;
        break;
      default:
        console.error('Type de trait invalide:', type);
        return null;
    }
    const index = options.findIndex(option => option.name.toLowerCase() === trait.toLowerCase());
    return index !== -1 ? index : null;
  }

  getTraitRarity(trait: string, type: string) {
    const index = this.getTraitIndex(trait, type);
    if (index === null) {
      console.error("Trait not found");
      return "";
    }
    if (type == "Transcendence" && index == 1) return "epic";
    if (index <= 2) return "legendary";
    if (index <= 5) return "epic";
    if (index <= 8) return "rare";
    return "";
  }

}

File: jar/src/main/java/com/foxyclan/jar/TraitOptionsService.java
package com.foxyclan.jar;

import java.util.Arrays;
import java.util.List;
import java.util.Map;

public class TraitOptionsService {

    private final List<String> headCoveringOptions = Arrays.asList(
        "Horns", 
        "Halo", 
        "Golden Tiara", 
        "Toxic Liquid", 
        "Robotic Mask", 
        "Engraved Helmet",
        "Golden Kabuto", 
        "Pilot Helmet",
        "Suit Hat", 
        "Crown", 
        "Ripple Of Magic",  
        "Straw Hat",
        "Ninja Bandana",
        "Headphones",
        "Bob",  
        "Earring" 
    ); //16

    private final List<String> eyesOptions = Arrays.asList(
        "Bloodshot", 
        "Scarlet", 
        "Golden Patterns", 
        "Toxic Green",
        "Robotic", 
        "Striped", 
        "Wisdom (Closed)",
        "Blindfolded", 
        "Sunglasses", 
        "Dollars", 
        "Sharingan",  
        "Happy (Closed)",
        "Bored",
        "Disdainful", 
        "Hypnotized",
        "Normal"
    ); //16

    private final List<String> mouthOptions = Arrays.asList(
        "Magma Fangs", 
        "Blue Psyche", 
        "Gold Mustache", 
        "Gas Mask", 
        "Robot Mask", 
        "Engraved Mask",
        "Gold Teeth",
        "Twig Nibbles",  
        "Smoking", 
        "Tongue Out", 
        "Stunned", 
        "Rabbit",
        "Hate", 
        "Happy",
        "Confused",
        "Neutral"
    ); //16

    private final List<String> clothesOptions = Arrays.asList(
        "Winged Dark Armor", 
        "Winged Golden Armor", 
        "Ornamental Armor", 
        "Military", 
        "Robot", 
        "Samurai",
        "Shoulder Cloak",
        "Pilot Suit",  
        "Suit", 
        "King's Cape",
        "Black Jacket",  
        "Overalls",
        "Blue Coat",  
        "Scarf", 
        "T-shirt",
        "Orange Jacket"
    ); //16

    private final List<String> furOptions = Arrays.asList(
        "Magma", 
        "Angel", 
        "Golden Obsidian", 
        "Dirty", 
        "Robot", 
        "Scarred", 
        "Beige",
        "Dark Brown",
        "Light Brown", 
        "Orange"
    ); //10

    private final List<String> backgroundOptions = Arrays.asList(
        "Hell", 
        "Heavenly Skies", 
        "Ancient Egypt", 
        "Green Gradient", 
        "Turquoise Gradient", 
        "Red Gradient", 
        "Soft Beige", 
        "Grey", 
        "Dark Blue", 
        "Red", 
        "Soft Mauve", 
        "Light Blue", 
        "Orange"
    ); //13

    private final List<String> transcendenceOptions = Arrays.asList(
        "Electrified",
        "Neon Nose"
    ); //1

    public String getTraitOption(String category, int index) {
        List<String> options = switch (category.toLowerCase()) {
            case "headcovering" -> headCoveringOptions;
            case "eyes" -> eyesOptions;
            case "mouth" -> mouthOptions;
            case "clothes" -> clothesOptions;
            case "fur" -> furOptions;
            case "background" -> backgroundOptions;
            case "transcendence" -> transcendenceOptions;
            default -> throw new IllegalArgumentException("Invalid category: " + category);
        };

        if (index < 0 || index >= options.size()) {
            throw new IndexOutOfBoundsException("Index out of range for category: " + category);
        }

        return options.get(index);
    }

    public int getTraitIndex(String category, String trait) {
        List<String> options = switch (category.toLowerCase()) {
            case "head covering" -> headCoveringOptions;
            case "eyes" -> eyesOptions;
            case "mouth" -> mouthOptions;
            case "clothes" -> clothesOptions;
            case "fur" -> furOptions;
            case "background" -> backgroundOptions;
            case "transcendence" -> transcendenceOptions;
            default -> throw new IllegalArgumentException("Invalid category: " + category);
        };

        int index = options.indexOf(trait);
        if (index == -1) {
            throw new IllegalArgumentException("Trait not found in category: " + category);
        }

        return index;
    }

    public String getBestTrait(String category, String value1, String value2) {
        if (value1 == null && value2 == null) return null;
        if (value1 == null) return String.format("%02d", this.getTraitIndex(category, value2));
        if (value2 == null) return String.format("%02d", this.getTraitIndex(category, value1));
        
        int index1 = this.getTraitIndex(category, value1);
        int index2 = this.getTraitIndex(category, value2);
        
        int bestIndex = Math.min(index1, index2);
        return String.format("%02d", bestIndex);
    }
    

    public String getTraitValue(Map<String, Object> metadata, String traitType) {
        Object attributesObj = metadata.get("attributes");
    
        if (attributesObj instanceof List<?>) {
            List<?> attributesList = (List<?>) attributesObj;
    
            for (Object attributeObj : attributesList) {
                if (attributeObj instanceof Map<?, ?> attributeMap) {
                    Object traitTypeObj = attributeMap.get("trait_type");
                    Object valueObj = attributeMap.get("value");
    
                    if (traitTypeObj instanceof String traitTypeStr && valueObj instanceof String valueStr) {
                        if (traitTypeStr.equals(traitType)) {
                            return valueStr;
                        }
                    }
                }
            }
        }
        return null;
    }



    private final Map<String, Integer> traitWeights = Map.ofEntries(
        // Background (poids 0)
        Map.entry("Background:00", 0), Map.entry("Background:01", 0), Map.entry("Background:02", 0),
        Map.entry("Background:03", 0), Map.entry("Background:04", 0), Map.entry("Background:05", 0),
        Map.entry("Background:06", 0), Map.entry("Background:07", 0), Map.entry("Background:08", 0),
        Map.entry("Background:09", 0), Map.entry("Background:10", 0), Map.entry("Background:11", 0),
        Map.entry("Background:12", 0),
        // Fur (poids 20)
        Map.entry("Fur:00", 20), Map.entry("Fur:01", 20), Map.entry("Fur:02", 20),
        Map.entry("Fur:03", 20), Map.entry("Fur:04", 20), Map.entry("Fur:05", 20),
        Map.entry("Fur:06", 20), Map.entry("Fur:07", 20), Map.entry("Fur:08", 20),
        Map.entry("Fur:09", 20),
        // Clothes (poids 40)
        Map.entry("Clothes:00", 40), Map.entry("Clothes:01", 40), Map.entry("Clothes:02", 40),
        Map.entry("Clothes:03", 40), Map.entry("Clothes:04", 40), Map.entry("Clothes:05", 40),
        Map.entry("Clothes:06", 40), Map.entry("Clothes:07", 40), Map.entry("Clothes:08", 40),
        Map.entry("Clothes:09", 40), Map.entry("Clothes:10", 40), Map.entry("Clothes:11", 40),
        Map.entry("Clothes:12", 40), Map.entry("Clothes:13", 40), Map.entry("Clothes:14", 40),
        Map.entry("Clothes:15", 40),
        // Eyes (poids 60)
        Map.entry("Eyes:00", 60), Map.entry("Eyes:01", 60), Map.entry("Eyes:02", 60),
        Map.entry("Eyes:03", 60), Map.entry("Eyes:04", 60), Map.entry("Eyes:05", 60),
        Map.entry("Eyes:06", 60), Map.entry("Eyes:07", 60), Map.entry("Eyes:08", 60),
        Map.entry("Eyes:09", 60), Map.entry("Eyes:10", 60), Map.entry("Eyes:11", 60),
        Map.entry("Eyes:12", 60), Map.entry("Eyes:13", 60), Map.entry("Eyes:14", 60),
        Map.entry("Eyes:15", 60),
        // Mouth (poids 80)
        Map.entry("Mouth:00", 80), Map.entry("Mouth:01", 80), Map.entry("Mouth:02", 140),
        Map.entry("Mouth:03", 120), Map.entry("Mouth:04", 120), Map.entry("Mouth:05", 120),
        Map.entry("Mouth:06", 80), Map.entry("Mouth:07", 140), Map.entry("Mouth:08", 180),
        Map.entry("Mouth:09", 80), Map.entry("Mouth:10", 80), Map.entry("Mouth:11", 80),
        Map.entry("Mouth:12", 80), Map.entry("Mouth:13", 80), Map.entry("Mouth:14", 80),
        Map.entry("Mouth:15", 80),
        // HeadCovering (poids 100)
        Map.entry("HeadCovering:00", 100), Map.entry("HeadCovering:01", 100), Map.entry("HeadCovering:02", 100),
        Map.entry("HeadCovering:03", 178), Map.entry("HeadCovering:04", 100), Map.entry("HeadCovering:05", 130),
        Map.entry("HeadCovering:06", 125), Map.entry("HeadCovering:07", 100), Map.entry("HeadCovering:08", 100),
        Map.entry("HeadCovering:09", 100), Map.entry("HeadCovering:10", 100), Map.entry("HeadCovering:11", 100),
        Map.entry("HeadCovering:12", 100), Map.entry("HeadCovering:13", 100), Map.entry("HeadCovering:14", 100),
        Map.entry("HeadCovering:15", 100),

        Map.entry("Transcendence:00", 179), Map.entry("Transcendence:01", 200)
    );

    public int getTraitWeight(String category, String code) {
        return traitWeights.getOrDefault(category + ":" + code, 0);
    }


}
